### PR TITLE
test(mc): #2242 Add unit test support for system-addon/

### PR DESCRIPTION
**Pull request by k88hudson**
This patch adds unit testing support for `.jsm`s in `system-addon/`.

It adds a karma configuration file and a `testmc` task (`tddmc` for continuous development), which run tests inside the folder `system-addon/test/unit/`.

It transforms `jsms` to use esmodule-style importing and exporting using `babel-plugin-jsm-to-esmodules`: see [this readme](https://github.com/k88hudson/babel-plugin-jsm-to-esmodules) for how it works. It has some limitations, but essentially it will rewrite `Cu.import` statements to `import` and `this.EXPORTED_SYMBOLS` to exports.

There is a basic test included for `ActivityStream.jsm`.

### CODE REVIEW

File: karma.mc.config.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
+const path = require("path");
+
+const CONFIG = {
```
Comment (dmose) [inline karma.mc.config.js]: Would a better name for this object be PATHS?

File: karma.mc.config.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  // The base directory of all source files (used for path resolution in webpack importing)
+  systemAddonDirectory: path.resolve(__dirname, "system-addon"),
+
+  // a RegEx matching all Cu.import statements of local files
+  resourcePathRegEx: /^resource:\/\/activity-stream\//
+};
+
+const preprocessors = {};
+preprocessors[CONFIG.testFilesPattern] = ["webpack"];
+module.exports = function(config) {
```
Comment (dmose) [inline karma.mc.config.js]: One line of whitespace above this for readability?

File: karma.mc.config.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  config.set({
+    singleRun: true,
+    browsers: ["Firefox"],
+    frameworks: ["mocha", "sinon", "chai"],
+    reporters: ["mocha"],
+    files: [CONFIG.testEntryFile],
+    preprocessors,
+    webpack: {
+      // This resolve config allows us to import with absolute paths, e.g. "lib/ActivityStream.jsm"
```
Comment (dmose) [inline karma.mc.config.js]: Maybe "with paths relative to the system-addon directory"?

File: karma.mc.config.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            exclude: [/node_modules/],
+            use: [{
+              loader: "babel-loader",
+              options: {plugins: [["jsm-to-esmodules", {basePath: CONFIG.resourcePathRegEx, replace: true}]]}
+            }]
+          }
+        ]
+      }
+    },
+    webpackMiddleware: {noInfo: true}
```
Comment (dmose) [inline karma.mc.config.js]: A comment describing noInfo would be convenient for readers.

File: system-addon/test/unit/lib/ActivityStream.test.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import {ActivityStream} from "lib/ActivityStream.jsm";
+
+describe("ActivityStream", () => {
+  it("should exist", () => {
+    assert.ok(ActivityStream);
+  });
+  it("should initialze with .initalized=false", () => {
```
Comment (dmose) [inline system-addon/test/unit/lib/ActivityStream.test.js]: nits: s/initialze/initialize/ and s/.inital/.initial/g

File: system-addon/test/unit/unit-entry.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+const req = require.context(".", true, /\.test\.js$/);
+const files = req.keys();
+
+// This exposes sinon assertions to chai.assert
+sinon.assert.expose(assert, {prefix: ""});
```
Comment (dmose) [inline system-addon/test/unit/unit-entry.js]: Why is this needed?

**Comment by k88hudson** [comment]
Yikes! What version of Chrome? 

Mine passes on node 7.6.0 and Chrome 57.0.2987 (Mac OS X 10.12.1)

File: system-addon/test/unit/unit-entry.js (hunk 6 above)
Comment (k88hudson) [inline system-addon/test/unit/unit-entry.js]: This is so we can do `assert.calledOnce`, etc., the same we do in our old mocha tests

File: system-addon/test/unit/lib/ActivityStream.test.js (hunk 5 above)
Comment (k88hudson) [inline system-addon/test/unit/lib/ActivityStream.test.js]: all those middle letters... do they really matter 😜

File: system-addon/test/unit/unit-entry.js (hunk 6 above)
Comment (dmose) [inline system-addon/test/unit/unit-entry.js]: Ah, indeed.  At some point, we might consider switching to https://github.com/domenic/sinon-chai for test readability.
